fileManager1API.js

```diff
@@ -0,0 +1,220 @@
+            if (k.startsWith(location)) {
+                for (let l of v) {
+                    ret.push(l);
+                }
+            }
+        }
+        return ret;
+    },
+
+    _onPropertyChanged: function(proxy, changed, invalidated) {
+        let property = changed.unpack();
+        if (property &&
+            ('XUbuntuOpenLocationsXids' in property ||
+             'OpenWindowsWithLocations' in property)) {
+            this._updateLocationMap();
+        }
+    },
+
+    _updateLocationMap: function() {
+        let properties = this._proxy.get_cached_property_names();
+        if (properties == null) {
+            // Nothing to check yet.
+            return;
+        }
+
+        if (properties.includes('OpenWindowsWithLocations')) {
+            this._updateFromPaths();
+        } else if (properties.includes('XUbuntuOpenLocationsXids')) {
+            this._updateFromXids();
+        }
+    },
+
+    _updateFromPaths: function() {
+        let pathToLocations = this._proxy.OpenWindowsWithLocations;
+        let pathToWindow = getPathToWindow();
+
+        let locationToWindow = new Map();
+        for (let path in pathToLocations) {
+            let locations = pathToLocations[path];
+            for (let i = 0; i < locations.length; i++) {
+                let l = locations[i];
+                // Use a set to deduplicate when a window has a
+                // location open in multiple tabs.
+                if (!locationToWindow.has(l)) {
+                    locationToWindow.set(l, new Set());
+                }
+                let window = pathToWindow.get(path);
+                if (window != null) {
+                    locationToWindow.get(l).add(window);
+                }
+            }
+        }
+        this._locationMap = locationToWindow;
+        this.emit('windows-changed');
+    },
+
+    _updateFromXids: function() {
+        let xidToLocations = this._proxy.XUbuntuOpenLocationsXids;
+        let xidToWindow = getXidToWindow();
+
+        let locationToWindow = new Map();
+        for (let xid in xidToLocations) {
+            let locations = xidToLocations[xid];
+            for (let i = 0; i < locations.length; i++) {
+                let l = locations[i];
+                // Use a set to deduplicate when a window has a
+                // location open in multiple tabs.
+                if (!locationToWindow.has(l)) {
+                    locationToWindow.set(l, new Set());
+                }
+                let window = xidToWindow.get(parseInt(xid));
+                if (window != null) {
+                    locationToWindow.get(l).add(window);
+                }
+            }
+        }
+        this._locationMap = locationToWindow;
+        this.emit('windows-changed');
+    }
+});
+Signals.addSignalMethods(FileManager1Client.prototype);
+
+/**
+ * Construct a map of gtk application window object paths to MetaWindows.
+ */
+function getPathToWindow() {
```

Comment:
I did this thinking about what would happen if you had multiple locations open. If I made getPathToWindow() take a path and search just that one window, then for multiple locations, it would have to repeat the search, and that looked more heavy-weight than searching through the map. So I figured, collect all the windows once and use that result to check each location.